### ISSUE TITLE
Reviewer Comments

### ISSUE THREAD

**Issue by Suvodeep90** [ai-se/GENERAL#2]
Comments for the Author:

All three reviewers think that the goals addressed in this paper is an important one to be addressed in SE research. Having said that they have several serious concerns. In no particular oder, the writing on the paper needs to be improved. Not just in style but in substance too. R1 and R2 see several key missing details. Next, all three reviewers have serious concerns about the experiments and the evaluation metrics. Finally, all three think that the evaluation itself by comparing to just a ZeroR model is not enough. The authors should compare against more state of the art models described by the reviewers and explain their choice. Besides these overlapping concerns, each reviewer has some specific concerns as well. For example, the defect ratio for the projects is atypical to the projects generally studied in other cross project defect prediction studies. And hence other concerns may not apply here. Are the good results a function of the data or the technique? Why not use
data from past studies? Given the substantial amount of work that still needs to be done, and that the reviewers were unable to judge several aspects accurately because of the missing details, I am recommending a reject.

Reviewer 1: *** Summary: ***
This paper presents a "fast" approach to identify a bellwether project among a large set of software projects. This bellwether project is then used to train a defect prediction model that will serve to detect defects in other projects. In other words, thanks to this bellwether project, this approach enables cross-project defect prediction. The "bellwether" concept is not new, however, the authors propose a new approach to identify a bellwether project, this approach being faster (i.e., lower complexity) than the state of the art approaches. Several experiments are performed. In particular, the authors check whether their approach is actually faster than the state of the art. 

*** Overall Comments ***
Defect prediction is an important topic in Software Engineering. However, the paper suffers from several major issues that I will classify under three categories:
1- The paper is not well-written. Additional readings would be welcomed to fix several typos, small issues, incorrect sentences,... The number of such issues is detrimental to the understanding of the paper. 
2- The contribution of this paper is limited. The empirical evaluation is limited too.   
3- Several related works are missing and there is no real comparison against these related works

I will now detail these three points:

*** Detailed Comments ***

** 1) Regarding the writing of the paper. **

There are several parts in the paper that are unclear. There are also several inconsistencies and typos. My feeling is that the paper was not ready for submission. 

- The paper is about defect prediction, if I am not mistaken, it is never explicitly mentioned whether the prediction is performed at the file, class, method, or statement level. Even if most of the state of the art approaches focus on file/class level prediction, it is better to explicitly say it. 

- Section 2.2 "How to Transfer Knowledge"
In this Section, you first explain that "There are 2 approaches frequently used in many research: similarity-based approaches and dimensional transforms."
And, you explain both approaches. 
And then, without any transition, you start a new paragraph "The Bellwether Method:" (p7 line 7). Please improve the transition. 

- p7 line 49: "When such similarities exist, then we exploit symmetries between different projects..." What do you mean by symmetries?   

- p8 line 6: "Accordingly, the rest of this paper performs the following experiment :"
Here you describe the steps of the GENERAL approach. So, you should not say "the rest of this paper", but rather something like: "the approach unfolds as follows". 

- p9 line 36: "we calculated 21 process metrics ... (see Table 2)". There are only 18 metrics in Table 2. 

- p9 line 41: "in Figure 2", Figure 2 is only presented in page 15!!! 6 pages later. Why?

- p11, line 3: "the use of RF (Random Forest) and SMOTE is endorsed by recent ICSE papers [1,15,65]."
[15] and [65] are not recent!

- Section "3.4 Implement Evaluation Criteria"
The description of the metrics is confusing. 
E.g., page 11 line 45 "After inspecting 20% LOC, we inspected m changes and found n defects."
Why 20%?
Recall: "This is the proportion of inspected defective changes"
What do you mean by "inspected". Is it not "detected"?
Pf: do you mean false positive?

- p14, RQ5, "This question was suggested to use by a colleague."
I do not understand this sentence. 

- p15, line 22, "Secondly we use Commit Guru [70] code to identify commits which have
bugs in them." You should better say: "commits that introduce a bug", or "commit inducing bugs"

- I found 4 pairs of duplicated citations. 
For instance, both [34] and [35] are actually the same citation.
It is also true for [45,46], [54,55], and [77,78].

- The overall number of projects in the dataset is not consistent: sometimes 758, sometimes 756 (and even 7580 on page 19, line 4). 

- some typos:
P4 line 41: "is that that the"
p7 line 46: "based on the writing of Devanbu et al.` [22] who" usually you should consider the first author: "Hindle et al. rather that Devandu et al.".
p10 line 43: "one if our motivation"  "s" is missing
p12 line 48 the following sentence is incorrect: "Combining numerous goals is often implemented as a multi-objective optimization predicate one model is better than another if it satisfies ..."
p13 line 15 "it it is"
p13 line 22 "Maximizevrecall and"

** 2) Regarding the contribution of the paper **
The contribution of this paper is limited. 
The core of this paper consists of using a hierarchical clustering approach to speed up the search for a bellwether project. Am I right? 
When the contribution is not "super impressive", the paper should be well-written, and/or the empirical evaluation has to be solid. 
In 1), I already explained why the paper is not well-written. 
I will now explain why the evaluation is not solid. 

First RQ2 "In theory, how much faster is GENERAL's hierarchical clustering?" should be removed. 
In the approach description, the authors already explained that the complexity of the algorithm is O(m*(N/m)^2).
There is no need to repeat this here. 
Moreover, in RQ2, which should focus on "in theory", the authors give empirical measures (that should be given in RQ3). 

Second, (less important) RQ1 is "How slow are conventional bellwether methods?"
This RQ is not well-posed. Indeed, you already presuppose the answer (i.e., the answer is already in the RQ: conventional bellwether methods are slow). An RQ should stay neutral. 

Third, Figure 6 (which is one of the most important figures of the paper) could be significantly improved. 
A lot of information is missing in the right part of the figure. 
Why not simply using traditional boxplots with clearly defined axis? 

Fourth, I am a big fan of the F1 metric which mixes precision and recall. 
Most state of the art approaches present F1. Why not here? 

Finally, I agree that a baseline is relevant, but ZeroR! 
This baseline is really not difficult to beat. Do you agree? 
Why not comparing against other state of the art approaches? 

***3) Related work **

First, as just said, there is no comparison with related work approaches in this paper. 

Second, the related work section is weak (actually I am not sure it is a real related work section). 
I simply searched for "cross project defect prediction" on Google Scholar and restrict the search with "Since 2019", and we can see that indeed several works are missing. For instance:
- A two-phase transfer learning model for cross-project defect prediction, IST 2019
- Improving defect prediction with deep forest, ÌST 2019
- Cross Project Defect Prediction via Balanced Distribution Adaptation Based Transfer Learning (JCST 2019)
etc...

(This one seems interesting but it has been published after the submission of this paper)
- Collective transfer learning for defect prediction, Neurcomputing 2020 

** 4) BTW Other question: **

Section 3.3 Data mining. 
The authors argue for feature selection. While I understand the need for feature selection when thousands or millions of features are used, it is less clear that with only 18 features, we still need feature selection. 

Reviewer 2: The paper proposes a new and improved method to automatically find bellwether projects from a large corpus of projects. Identifying these bellwether projects are useful in practice as they provide insights that are stable across multiple projects. In this paper, the authors propose a new method call GENERAL which first performs a feature selection and then hierarchically clusters the datasets. Treating the different stages of the hierarchical cluster as a tree they find a bellwether project at each level which they then push up to the root. This project is finally presented as the bellwether project. Such an approach improves the run time over the existing approach dramatically and thereby makes adoption of GENERAL more easy.

Pros: 
1.      The paper tackles an extremely important problem in defect prediction. Conclusion instability has been a constant issue in defect prediction. However, most solutions till date have not been easy to adopt in practice. The proposed method seems promising given the the runtimes on such vast corpus seems optimal. 
2.      The paper attempts to not only show transferable performance, but also seeks to explain the factors that might be responsible for bugs.
3.      The studied RQs are relevant and the findings are actionable. 

Cons: 
While the goal of the paper is extremely laudable there are several aspects of the paper that are unclear and some serious concerns with the methodology and the conducted experiments. My major concerns with the paper are that (1. The experiment setup at various points is not clear (comment 2, comment 4, comment 10), 2. Weak baselines (comment 4) 3. Lack of clarity around how feature importances are computed (comment 10, comment 11, comment 12). These concerns lead me to only believe that at best, the authors find that GENERAL is a more robust method than the methods used identify Bellwether projects earlier. Therefore, for the other findings to come to fore, I would like the authors to clarify the major concerns. 

I outline all concerns below which I would like addressed before this paper can be considered for publication. I outline my concerns in each section of the paper as follows. 
Introduction: 
1.      The authors seem to overclaim that they can authoritatively assert across hundreds of project that can establish what factors do and do not contribute to the defects happening. However, in their RQ6 and Discussion section, they only have frequency based feature importance that doesn't necessarily hold for all the studied projects. Therefore, I would recommend the authors to tone down their findings to appropriately. 
About GENERAL: 
I have several confusions here.
2.      The authors use CFS to prune conduct a feature selection. However CFS could typically select different subset of features for different datasets. For instance in Dataset(a) the selected features could f1,f2,f3 and in Dataset(b) could yield f4,f5. However from the documentation of Brich (https://scikit-learn.org/stable/modules/generated/sklearn.cluster.Birch.html) the arrays that need to submitted to BRICH needs to be of similar shape. So I am wondering how the different features being selected is not a problem for the subsequent clustering. Please clarify how CFS to BRICH works with an example in the paper as I think it will greatly enhance readability 
3.      Several of the author's prior work argue that Hyperparameter tuning of the dataminers used in a study could have a significant impact on the observed results. Therefore, I wonder why the random forest learner used as a part of the GENERAL pipeline was not hyperparameter tuned. I wonder, if the learner's hyper parameters are tuned, would the GENERAL method produce more competitive Bellwether. Please explain the rationale for not tuning the hyperparameters. 
Experimental Methods
4.      Experimental Setup: The experimental setup, in particular the test 2 and train 2 is extremely confusing and I would urge the authors to explain it with an example. For instance, from my reading of the paper, train 1 is 756 projects and test 1 is 76 projects (which is clearly stated in the paper). So this test_2 which is 76 projects is split using a 2:1 split. For instance approximately 50 projects make up train_2 and 26 projects make up test_2? 
a.      Therefore, when the self learner is trained how is this train train_2 leveraged? The authors in the paper states that 90% of training set (which we call train_2) and a 10% test set (which we call test_2). Now isn't train_2 and test_2 comprised of 50 projects and  26 projects respectively. So for the self are these project data merged together to form the train and test set (if this is the case, it does not make sense to me as its not particularly local learning anymore). Or each project from test_2 (76 projects) undergo a 90:10 split (which makes more sense to me). This aspect is extremely confusing. Because of this, the way the trian_2 and test_2 is used for the rest of the setup is also not very clear to me. Also if each project is split into 90:10 how are the results aggregated? I would like the authors to make how these datasets in train_2 and test_2 used and how the performance measures are being computed on them clear as it direwctly relates to the validity of the
findings 
b.      Using ZeroR as the baseline. I understand that authors justify this choice by stating that ZeroR is a simple method. However, this method feels more like a strawman method. I suggest that the authors should use a more competitive baseline (e.g., a random forest or a Fast Frugal Tree (from one of the author's previous paper') as both of them are quite simple). 
c.      Which learner is used for the Global method?
d.      Why compare results on different levels? 
5.      The experimental setup does not elaborate how the feature importance which are used in RQ6 are computed

Results: 
6.      Minor error: Pg 19, line 4: how is it 7580 projects? 
7.      Minor error: Pg21, line 37 media  median
8.      Minor error: Pg22 Gig 6: Instead of General_level the table calls them Bubble_level
9.      From the result of RQ5, the authors state that learning from dozens of projects gives the same performance. Then do we even need to construct a cluster?  Wouldn't just seeing the best bellwethers are the level 2 enough? Won't this dramatically cut down the computation time? Please remark and add more discussion around why GENERAL should still need level 1 and level 2
10.     RQ6: I do not understand how the feature importance are computed for this Research question. Are the bellwether models interpretated with a method like gini importance?  How are the features marked in RED ascertained important? Earlier in the paper, the authors state that they use CFS to do this, however as I mentioned in my earlier comment (comment 5), different datasets would have different features. In that context how are the feature importances calculated. Such confusion, does not allow me to interpret what exactly these important features mean. Please explain how the feature importance (in particular global importances) are computed. 
11.     Shouldn't the feature importance be derived based on interpreting the model? Rather than using feature selection (which only maximizes the correlation between independent and dependent factors)? Therefore, it doesn't really remark anything about the factors that the bellwether model thinks lead to a bug, rather only what factors are mostly correlated with a bug. (If that's what authors want to do that's fine. But it should be clearly stated)
12.     Finally, if the model is indeed interpreted, then it should be done so without SMOTE (http://chakkrit.com/assets/papers/tantithamthavorn2018imbalance.pdf) as previous study finds class rebalancing might affect the interepretation of machine learning models.
Discussion:
13.     Same issue that I mention in comment 13 impacts the discussion section too. I am not certain how the feature importances are calculated. 

Reviewer 3: This paper presents an interesting idea to build generalized defect prediction models. The authors propose a fast approach to identify a bellwether project from a large set of projects. The results show that the defect prediction model built from the bellwether model can be generalized. The generalized model gives insights on defect reasoning.

Strength:
+ interesting topic with considerable impact
+ thoughtful choice of most methods and awareness of threats to validity
+ a large scale experiment

Weakness:
- findings from the GENERAL model have been applied in modern development, and do not bring new insights
- lack of comparision with state-of-the-art generalized models
- only Java projects with relatively high defect ratio

This paper addresses an interesting topic, generalizing defect prediction model. In recent years, more studies try to generalize defect prediction models using various ways. A generalized model not only helps give important advices to developers, but make it easier for developers to apply defect prediction models in practice.

The main idea is find a bellwether project from a larget set of projects. The global bellwether project is selected from a subset of bellwether projects via hierarchical clustering. In a leaf cluster, a project is the bellwether project if model built from it performs the best (measured using a combination of five performance metrics). Therefore, model built from the global bellwether project performs the best among all bellwether models.

Although the paper is well written and methods are carefully chosen. There are major issues to be addressed:

1. Findinds from the global bellwether model have been applied in practice, and do not bring new insights.
- Code churn has already been proven as a good defect indicator in many previous studies. For example, 
  J. C. Munson and S. G. Elbaum, "Code churn: a measure for estimating the impact of code change," Proceedings. International Conference on Software Maintenance (Cat. No. 98CB36272), Bethesda, MD, USA, 1998, pp. 24-31, doi: 10.1109/ICSM.1998.738486.
  Do more changes mean higher quality? A more often changed file is examined more often too.
- Mimimizing the use of amateurs. In nowdays development, project development has a pull request mechanism to maintain code quality. Amateur developers must submit a pull request which is reviewed by experienced developers.
- It is true to have a clear owner of each file and let others who are unfaimilar with the file touch it. In practice, each file often has a clear owner and overlapping editing is highly avoided. This is because resolving merge conflicts is headache and problematic. Developers always try to avoid commit conflicts. A nature way is to seperate files by devlopers.
- Making changes in multiple places across a file. In modern software development, simplifing methods/files has been a common practice. FYI., check the following articles from developers.
[Code Refactoring Best Practices: When (and When Not) to Do It \| AltexSoft](https://www.altexsoft.com/blog/engineering/code-refactoring-best-practices-when-and-when-not-to-do-it/)
[Small Files Are Your Friends - Codecraft](https://codecraft.co/2013/03/21/small-files-are-your-friends/)
[code organization - Breaking up classes and methods into smaller units - Software Engineering Stack Exchange](https://softwareengineering.stackexchange.com/questions/244688/breaking-up-classes-and-methods-into-smaller-units)

2. Experiment design
- Defects are mined by processing commit messages, so more changes, more commit messages, more likely to have 'bugs'. In addition, a bug may be fixed in multiple commits with same commit message. More careful examination needs to be performed.
- Defect ratio is relatively too high in the 756 projects. In software industry, a common defect ratio is around 20%.
- As process metrics are general across programming languages, it is highly recommended to include projects developed using other languages, such as c/c++, object c, python, nodejs.
- ifa_auc seems to be problematic, as it cannot distinguish ZeroR and other models.
- precion/recall/pf is highly impacted by the choice of threshold. Why not use AUC that has been widely used in defect prediction studies?
- F-measure is a balance of precision and recall, it is also interesting to report F-meausure value.
- A comparison with other generalized model is needed. For example,
  F. Wu et al., "Cross-Project and Within-Project Semisupervised Software Defect Prediction: A Unified Approach," in IEEE Transactions on Reliability, vol. 67, no. 2, pp. 581-597, June 2018.
  F. Zhang, Q. Zheng, Y. Zou and A. E. Hassan, "Cross-project defect prediction using a connectivity-based unsupervised classifier", Proc. 38th Int. Conf. Softw. Eng., pp. 309-320, 2016.

3. Inaccurate statements
- P23, L43, the conclusion is inaccurate. Standard bellwether model outperfomrs GENERAL bellwether model in terms of Popt20 as show in Fig. 6.
- P27, L6. 'only a small subset' is misleading. As shown in Fig. 9, among the top 11 most frequent metrics, 5 appears only in local models, is similar to the amount (6) appearing in global models.

4. Other issues:
- Global insights are useful, but a development team is more interesting on what model works for its own project. A further thinking on applying generalized defect prediction models in practice is needed. As a result, the performance of the generalized defect prediction should work as good as 'self' models.
- It is more interesing to dig deeper in the global bellwether project, why model built from it works for all other projects?
- Are ifa_auc values presented in Fig.6 correct?

Minor issues:

- Should models be named as "GENERAL_level*" in Fig. 6?
- typos: 
  Pg13, L13: another it
  Pg13, L22: Maximizevrecall
  P17, L7, learner build
